storm/src/main/storm/mesos/schedulers/DefaultScheduler.java

```diff
@@ -0,0 +1,338 @@
+    for (Protos.Offer offer : offers.newestValues()) {
+      boolean isOfferUseful = false;
+      for (String currentTopology : topologiesMissingAssignments) {
+        boolean supervisorExists = supervisorExists(offer.getHostname(), existingSupervisors, currentTopology);
+        TopologyDetails topologyDetails = topologies.getById(currentTopology);
+        if (isFit(offer, topologyDetails, supervisorExists)) {
+          isOfferUseful = true;
+          break;
+        }
+      }
+      if (!isOfferUseful) {
+        log.info("Declining Offer " + getTrimmedString(offer.getId()) + " because it does not fit any of the topologies that need assignments");
+        offers.clearKey(offer.getId());
+      }
+    }
+
+    List<WorkerSlot> allSlots = new ArrayList<>();
+
+    Map<String, List<OfferResources>> offerResourcesListPerNode = getOfferResourcesListPerNode(offers);
+
+    for (String currentTopology : topologiesMissingAssignments) {
+      TopologyDetails topologyDetails = topologies.getById(currentTopology);
+      int slotsNeeded = topologyDetails.getNumWorkers();
+
+      double requestedWorkerCpu = MesosCommon.topologyWorkerCpu(mesosStormConf, topologyDetails);
+      double requestedWorkerMem = MesosCommon.topologyWorkerMem(mesosStormConf, topologyDetails);
+
+      log.info("Trying to find " + slotsNeeded + " slots for " + topologyDetails.getName());
+      if (slotsNeeded == 0) {
+        continue;
+      }
+
+      String[] nodes = new String[offerResourcesListPerNode.keySet().size()];
+      int i = 0;
+      for (String node : offerResourcesListPerNode.keySet()) {
+        nodes[i++] = node;
+      }
+
+      boolean slotFound;
+
+      do {
+        slotFound = false;
+        for (Object currentNode : nodes) {
+          boolean supervisorExists = supervisorExists((String) currentNode, existingSupervisors, currentTopology);
+
+          for (OfferResources resources : offerResourcesListPerNode.get(currentNode)) {
+            boolean isFit = this.isFit(resources, topologyDetails, supervisorExists);
+            if (isFit) {
+              log.info(resources.toString() + " is a fit for " +
+                       topologyDetails.getId() + " requestedWorkerCpu: " + requestedWorkerCpu + " requestedWorkerMem: " +
+                       requestedWorkerMem);
+              MesosWorkerSlot mesosWorkerSlot = createWorkerSlotFromOfferResources(resources, topologyDetails, supervisorExists);
+              if (mesosWorkerSlot == null) {
+                continue;
+              }
+              String slotId = mesosWorkerSlot.getNodeId() + ":" + mesosWorkerSlot.getPort();
+              mesosWorkerSlotMap.put(slotId, mesosWorkerSlot);
+              // Place this offer in the first bucket of the RotatingMap so that it is less likely to get rotated out
+              offers.put(resources.getOfferId(), resources.getOffer());
+              allSlots.add(mesosWorkerSlot);
+              if (--slotsNeeded == 0) {
+                break;
```

Comment:
As we found this afternoon, this should be a "double break", where we check for `slotsNeeded == 0` before continuing in the `for (Object currentNode : nodes) {` loop.   Otherwise we were setting aside too many slots for a given topology to use under certain circumstances, which resulted in an uneven spread of tasks across workers.